enhancements/ocp-coreos-layering.md

```diff
@@ -0,0 +1,496 @@
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Change RHEL CoreOS as shipped in OpenShift to be a "base image" that can be used as in layered container builds and then booted.  This will allow custom 3rd party agents delivered via RPMs installed in a container build.  The MCO will roll out and monitor these custom builds the same way it does for the "pristine" CoreOS image today.
+
+This is the OpenShift integration of [ostree native containers](https://fedoraproject.org/wiki/Changes/OstreeNativeContainer) or [CoreOS layering](https://github.com/coreos/enhancements/pull/7) via the MCO.
+
+## Motivation
+
+1. We want to support more customization, such as 3rd party security agents (often in RPM format).
+1. It is very difficult today to roll out a hotfix (kernel or userspace).
+1. It should be possible to ship configuration with the same mechanism as adding content.
+1. There are upgrade problems today caused by configuration updates of the MCO applied separately from OS changes.
+1. It is difficult to inspect and test in advance what configuration will be created by the MCO without having it render the config and upgrade a node.
+
+### Goals
+
+- [ ] Administrators can add custom code alongside configuration via an opinionated, tested declarative build system and via e.g. `Dockerfile` as an unvalidated generic mechanism
+- [ ] The output is a container that can be pushed to a registry, inspected, and processed via security scanners. 
+- [ ] Further, it can be executed as a regular container (via podman or as a pod) *and* booted inside or outside of a cluster
+- [ ] Gain [transactional configuration+code changes](https://github.com/openshift/machine-config-operator/issues/1190) by having configuration+encode captured atomically in a container
+- [ ] Avoid breaking existing workflow via MachineConfig (including extensions)
+- [ ] Avoid overwriting existing custom modifications (such as files managed by other operators) during upgrades
+
+### Non-Goals
+
+- While the base CoreOS layer/ostree-container tools will be usable outside of OpenShift, this enhancement does not cover or propose any in-cluster functionality for exporting the forward-generated image outside of an OpenShift cluster. In other words, it is not intended to be booted (`$ rpm-ostree rebase <image>`) from outside of a cluster.
+- This proposal does not cover generating updated "bootimages"; see https://github.com/openshift/enhancements/pull/201
+- Don't change existing workflow for RHEL worker nodes
+
+## Proposal
+
+**NOTE: Nothing in this proposal should be viewed as final.  It is highly likely that details will change.  It is quite possible that larger architectural changes will be made as well.**
+
+1. The `machine-os-content` shipped as part of the release payload will change format to the new "native ostree-container" format, in which the OS content appears as any other OCI/Docker container.  
```

Comment:
@control-d I think this means we can now ship this image on registry.redhat.io (or the container catalog) and not just on Quay.

Reply:
I wonder if this does or would solve some of the CVE reporting issues we have with RHCOS today?

Reply:
Yes, a big sub-goal here is that the container image (though it has some "ostree stuff" in there) otherwise looks like e.g. the UBI base image, and should be understandable to security scanners like Clair directly.

There's some details though around how we handle extensions - those would not be visible to scanners.